### ISSUE TITLE
SEC-17: Hostname verification on SSL configuration

### ISSUE THREAD

**Issue by acdemyers** [OpenLiberty/open-liberty#7496]
Feature to add hostname verification configurable at the SSL configuration level.

**Comment by acdemyers** [comment]
Aha feature can be found:  https://bigblue.aha.io/features/SEC-17

**Comment by acdemyers** [comment]
## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A. AHA Feature marked.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A. AHA Feature marked.
- [ ] Complete any To-Do's from the POC Review, and create a new TO-DO in the AHA Feature for the "Design Approval Request" and assign it to Ian Robinson. Leave the date blank. (Feature Owner)
- [ ] Design / WAD Approval (Ian Robinson) or N/A. AHA Feature marked.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A. AHA Feature marked.
- [ ] AHA Feature has link to Github epic or issue. Github epic has link back to AHA Feature. (Feature Owner or Backlog Subtribe PM).
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary"
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approval or No Design Approval on the AHA Feature.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by atosak** [comment]
ID issue: https://github.ibm.com/was-liberty/liberty-docs/issues/1335

**Comment by acdemyers** [comment]
FVT issue: https://github.ibm.com/was-liberty/WS-CD-Open/issues/17031

**Comment by mtamboli** [comment]
SVT is not planning explicit testing for this and this testing for this feature is more functional. We are trying to add this SSL property for some of our outbound SSL connection in regression NEST testing. This NEST testing is tracked by  https://github.ibm.com/was-svt/NEST-HostingEnv/issues/284

**Comment by marikaj123** [comment]
@acdemyers - Does the feature need translation?

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by acdemyers** [comment]
> L2 has requested STE slides for this feature. The STE template can be found at the link below:
> 
> https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm
> 
> Please upload the completed slides to the same STE Archive BOX folder. Thanks!

**Comment by acdemyers** [comment]
@donbourne 
Serviceability Approval Comment :
	1	WAD — This feature was approved for now no WAD needed
	2	a). Hostname verification is server identity check performed by the JDK.   The SSL Handshake errors if the hostname used to access a server does not match the information in the server’s certificate.  SSL error tend to get buried and bubble in a very generic form.  But this error was one I was able to wrap in the call to the trust manager.  So if JDK fails a hostname verification check in the Liberty will issue a message to the logs: 

          CWPKI0824E: SSL HANDSHAKE FAILURE:  Host name verification error while connecting to host [localhost].  The host name used to access the server does not match the server certificate's SubjectDN or Subject Alternative Name information.  The extended error message from the SSL handshake exception is: [No subject alternative DNS name matching localhost found].
		
          b) This was demo-ed on the end of iteration call on 6/4/2019.   I showed the new error and that it shows up in the message log.
          c) The new message was approved by L2 and ID.  


	3.   SVT added this comment to the feature:

		SVT is not planning explicit testing for this and this testing for this feature is more functional. We are trying to add this SSL property for some of our outbound SSL connection in regression NEST testing. This NEST testing is tracked by https://github.ibm.com/was-svt/NEST-HostingEnv/issues/284

	4	The security L2/L3 queues will hand support of this feature.   For L2 WASSEC queue and L3 the WL3SEC queue.  The L3 team is aware due to the team being a part of our development team.   L2 has been informed and I’m waiting on a education session to be setup with them.

**Comment by acdemyers** [comment]
@skasund The STE was placed in box https://ibm.ent.box.com/file/469844628313

**Comment by skasund** [comment]
@acdemyers  Thanks for the slides!  I've provided the STE approval.

**Comment by gscottj** [comment]
This feature does not appear to add any new or change any existing user interfaces in a manner that requires accessibility verification testing.

**Comment by marikaj123** [comment]
@arkarkala - the feature need to have focal point approvals for Translation, ID, Performance and SVT.

**Comment by mtamboli** [comment]
SVT did not plan on explicit testing for this feature. Some Implicit testing is getting added to NEST. We are depending on the FVT for the test coverage for this feature. I am approving this feature.

**Comment by marikaj123** [comment]
@acdemyers and @arkarkala - Checked with the team and the returned Emergency files are waiting to get into Integration.  The team is currently working on it.

**Comment by sabolo** [comment]
The cross platform test run over the weekend was impacted by a bad code delivery that has now been reverted.  We are going to do a smaller run mid week and verify that the tests are passing.  Assuming all passes then we should be good to sign off the FAT work here.

**Comment by arkarkala** [comment]
@marikaj123 - The translated files are in the integration branch. I think it is ready for Globalization approval.

**Comment by marikaj123** [comment]
@arkarkala they need to exist in a green Release build before I can approve.

**Comment by arkarkala** [comment]
@marikaj123 - The translated files are in the master branch. I think it is ready for Globalization approval.

**Comment by marikaj123** [comment]
@dave-waddling - Is the feature ready to receive the FAt approval?

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Apologies for the delay in getting this feature reviewed, we had a couple of issues getting 19.0.0.6 out of the door which our squad were focusing on. I've finished reviewing the Feature Test summary (#7702) and have added the FAT Focal Approval to this feature.

**Comment by marikaj123** [comment]
@acdemyers - Did you provide a blog for this feature in 19006?

**Comment by marikaj123** [comment]
For 19006, since the feature received all focal point approvals and the blog was published then it can be closed.

**Comment by acdemyers** [comment]
All approvals have been provided.   Closing the epic.
